pkg/distributor/distributor.go

```diff
@@ -1192,6 +1203,26 @@ func (d *Distributor) metricsMiddleware(next PushFunc) PushFunc {
 	}
 }
 
+// outerMaybeDelayMiddleware is a middleware that may delay ingestion if configured.
+func (d *Distributor) outerMaybeDelayMiddleware(next PushFunc) PushFunc {
+	return func(ctx context.Context, pushReq *Request) error {
+		start := d.now()
+		// Execute the whole middleware chain.
+		err := next(ctx, pushReq)
+
+		userID, userErr := tenant.TenantID(ctx) // Log tenant ID if available.
+		if userErr == nil {
+			// Target delay - time spent processing the middleware chain including the push.
+			// If the request took longer than the target delay, we don't delay at all as sleep will return immediately for a negative value.
+			d.sleep(d.limits.DistributorArtificialIngestionDelay(userID) - d.now().Sub(start))
```

Comment:
I would recommend applying a jitter to be more realistic. You can do it like this:

```
if delay := d.limits.DistributorArtificialIngestionDelay(userID) - d.now().Sub(start); delay > 0 {
  d.sleep(util.DurationWithJitter(delay, 0.1))
}
```

The code above applies a 10% jitter (both positive and negative). It must take in input positive values, reason why I'm checking delay > 0.

Reply:
Done in https://github.com/grafana/mimir/pull/10107/commits/945dfb844564683d41a87fd4f290e626007805cd